libcloud/common/aws.py

```diff
@@ -178,6 +180,68 @@ def _get_aws_auth_param(self, params, secret_key, path='/'):
         return b64_hmac.decode('utf-8')
 
 
+class V4SignedAWSConnection(AWSTokenConnection):
+
+    def add_default_params(self, params):
+        params['Version'] = self.version
+        return params
+
+    def pre_connect_hook(self, params, headers):
+        now = datetime.utcnow()
+        headers['X-AMZ-Date'] = now.strftime('%Y%m%dT%H%M%SZ')
+        headers['Authorization'] = self._get_authorization_v4_header(params, headers, now)
+
+        return params, headers
+
+    def _get_authorization_v4_header(self, params, headers, dt):
+        # TODO: according to AWS spec (and RFC 2616 Section 4.2.) excess whitespace
+        # from inside non-quoted strings should be stripped. Now we only strip the
+        # start and end of the string. See http://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html
+        canonical_headers = '\n'.join([':'.join([k.lower(), v.strip()])
+                                       for k, v in sorted(headers.items())])
+        canonical_headers += '\n'
+
+        signed_headers = ';'.join([k.lower() for k in sorted(headers.keys())])
+
+        # For self.method == GET
+        request_params = '&'.join(["%s=%s" % (urlquote(k, safe=''), urlquote(v, safe='-_~'))
+                                   for k, v in sorted(params.items())])
+        payload_hash = hashlib.sha256('').hexdigest()
```

Comment:
Hm, is this correct? It looks like it's just calculating SHA256 hash of an empty string.